### ISSUE TITLE
Chapter 4 Accumulators example doesn't trigger TCO.

### ISSUE THREAD

**Issue by meowjesty** [purescript-contrib/purescript-book#41]
Hi, the [example](https://github.com/dwhitney/purescript-book/blob/master/text/chapter4.md) on how purescript will turn tail recursive functions into js `while` loops doesn't actually work as expected.

Code like this:

```purescript
length_tail :: forall a. Array a -> Int
length_tail arr = len_ arr 0
    where
        len_ :: forall a. Array a -> Int -> Int
        len_ arr acc =
            if null arr then
                0
            else
                len_ (fromMaybe [] $ tail arr) acc + 1
```

Will become your stack overflowing call in javascript:

```js
var length_tail = function (arr) {
    var len_ = function (arr1) {
        return function (acc) {
            var $10 = Data_Array["null"](arr1);
            if ($10) {
                return 0;
            };
            return len_(Data_Maybe.fromMaybe([  ])(Data_Array.tail(arr1)))(acc) + 1 | 0;
        };
    };
    return len_(arr)(0);
};
```

You need to change `acc + 1` into `(acc + 1)`:

```purescript
length_tail :: forall a. Array a -> Int
length_tail arr = len_ arr 0
    where
        len_ :: forall a. Array a -> Int -> Int
        len_ arr acc =
            if null arr then
                0
            else
                len_ (fromMaybe [] $ tail arr) (acc + 1)
```

To get the `while` loop version:

```js
var length_tail = function (arr) {
    var len_ = function ($copy_arr1) {
        return function ($copy_acc) {
            var $tco_var_arr1 = $copy_arr1;
            var $tco_done = false;
            var $tco_result;
            function $tco_loop(arr1, acc) {
                var $10 = Data_Array["null"](arr1);
                if ($10) {
                    $tco_done = true;
                    return 0;
                };
                $tco_var_arr1 = Data_Maybe.fromMaybe([  ])(Data_Array.tail(arr1));
                $copy_acc = acc + 1 | 0;
                return;
            };
            while (!$tco_done) {
                $tco_result = $tco_loop($tco_var_arr1, $copy_acc);
            };
            return $tco_result;
        };
    };
    return len_(arr)(0);
};
```

I just started learning purescript, so sorry if I'm missing something here.

This code was written in a project using:
- `spago`: v0.10.0.0
- `purs`: v0.13.4

**Comment by JamieBallingall** [comment]
So the example has been updated for better style and to remove warnings. However, this issue still occurs and the parenthesis still fix it. While this is a simple fix, I feel like it would be nice to provide an explanation. Sadly, I don't myself understand it.

**Comment by thomashoneyman** [comment]
The example isn't correct; it needs the parentheses because of precedence rules (the function is being applied to `acc` _and then_ we add one to the result). 

This code:

```purs
length' (fromMaybe [] $ tail arr) acc + 1
```

is parsed like this:

```purs
(length' (fromMaybe [] $ tail arr) acc) + 1
```

instead of the way it's intended from my reading of the example, which is:

```purs
length' (fromMaybe [] $ tail arr) (acc + 1)
```

Just to check this, I threw it into try.purescript.org and looked at the output, and indeed the version without parenthesis emits this JS:

```js
return length$prime(Data_Maybe.fromMaybe([  ])(Data_Array.tail(arr$prime)))(acc) + 1 | 0;
```

which is adding 1 to the result of `length'` after it is called on `acc`, which isn't going to fire the TCO (or increment the accumulator).

So I'd suggest just fixing the example without further explanation -- it's not that this is a tricky case, it's just that the example is incorrect.

**Comment by milesfrain** [comment]
Thanks a bunch for reporting this and providing an explanation of what's happening.
PR to fix in #97
